File: solution/Python/shared/providers/configurable.py
"""Base classes for configurable machines."""
from abc import ABC, abstractmethod
from enum import Enum
from typing import Any, Optional, Union
from dataclasses import dataclass


class SettingType(str, Enum):
    """Type of setting value."""
    STRING = "string"
    NUMBER = "number"
    BOOLEAN = "boolean"


class UnitOfMeasure(str, Enum):
    """Unit of measure for numeric settings."""
    DEGREE_CELSIUS = "°C"
    DEGREE_FAHRENHEIT = "°F"
    BAR = "bar"
    PSI = "psi"
    RPM = "rpm"
    RPS = "rps"
    SECOND = "s"
    MINUTE = "min"


@dataclass
class SettingBase:
    """Base class for all settings."""
    namespace: Optional[str]
    identifier: str
    description: str
    nullable: Optional[bool] = None


@dataclass
class NumericSetting(SettingBase):
    """Numeric setting with optional unit of measure and range."""
    type: SettingType = SettingType.NUMBER
    default_value: Optional[float] = None
    uom: Optional[UnitOfMeasure] = None
    min_value: Optional[float] = None
    max_value: Optional[float] = None


@dataclass
class StringSetting(SettingBase):
    """String setting."""
    type: SettingType = SettingType.STRING
    default_value: Optional[str] = None


@dataclass
class BooleanSetting(SettingBase):
    """Boolean setting."""
    type: SettingType = SettingType.BOOLEAN
    default_value: Optional[bool] = None


Setting = Union[NumericSetting, StringSetting, BooleanSetting]


@dataclass
class SettingValue:
    """A setting value to be applied."""
    identifier: str
    value: Optional[Union[str, float, bool]] = None
    uom: Optional[UnitOfMeasure] = None


@dataclass
class SettingError:
    """An error in a setting value."""
    identifier: str
    message: str


class ConfigurableMachine(ABC):
    """Abstract base class for configurable machines."""
    
    @property
    @abstractmethod
    def settings(self) -> list[Setting]:
        """Get the list of settings for this machine."""
        pass
    
    def _verify_settings(self, settings: list[SettingValue]) -> list[SettingError]:
        """Verify that settings are valid.
        
        Performs the following checks:
        * Each identifier must exist in settings (see getter)
        * No duplicate identifiers
        * There must be a value or a default value for each setting that is not nullable
        * Value must be of the correct type
        * If numeric, value must be >= minValue and <= maxValue
        * If numeric, uom must fit or be convertible to the setting's uom
        
        Args:
            settings: List of settings to verify
            
        Returns:
            List of errors (empty if valid)
        """
        errors: list[SettingError] = []
        
        # Build definition map
        definition_map: dict[str, Setting] = {}
        for definition in self.settings:
            definition_map[definition.identifier] = definition
        
        # Check for duplicates
        seen_provided: set[str] = set()
        duplicate_provided: set[str] = set()
        for provided in settings:
            if provided.identifier in seen_provided:
                duplicate_provided.add(provided.identifier)
            else:
                seen_provided.add(provided.identifier)
        
        for identifier in duplicate_provided:
            errors.append(SettingError(
                identifier=identifier,
                message="Duplicate setting identifier provided"
            ))
        
        # Build provided map (keep first occurrence)
        provided_map: dict[str, SettingValue] = {}
        for provided in settings:
            if provided.identifier not in definition_map:
                errors.append(SettingError(
                    identifier=provided.identifier,
                    message="Unknown setting identifier"
                ))
                continue
            
            if provided.identifier not in provided_map:
                provided_map[provided.identifier] = provided
        
        # Verify each definition
        for identifier, definition in definition_map.items():
            provided = provided_map.get(identifier)
            nullable = definition.nullable is True
            
            has_provided_value = provided is not None and provided.value is not None
            default_value = getattr(definition, 'default_value', None)
            has_default_value = default_value is not None
            
            if not nullable and not has_provided_value and not has_default_value:
                errors.append(SettingError(
                    identifier=identifier,
                    message="Missing value (no default and not nullable)"
                ))
                continue
            
            # Determine effective value
            effective_value = provided.value if has_provided_value else default_value
            if effective_value is None:
                # Nullable without value is fine
                if definition.type != SettingType.NUMBER and provided is not None and provided.uom is not None:
                    errors.append(SettingError(
                        identifier=identifier,
                        message="Unit of measure is only allowed for numeric settings"
                    ))
                continue
            
            # Check if UOM is used on non-numeric settings
            if definition.type != SettingType.NUMBER and provided is not None and provided.uom is not None:
                errors.append(SettingError(
                    identifier=identifier,
                    message="Unit of measure is only allowed for numeric settings"
                ))
            
            # Type-specific validation
            if definition.type == SettingType.STRING:
                if not isinstance(effective_value, str):
                    errors.append(SettingError(
                        identifier=identifier,
                        message="Value must be a string"
                    ))
            elif definition.type == SettingType.BOOLEAN:
                if not isinstance(effective_value, bool):
                    errors.append(SettingError(
                        identifier=identifier,
                        message="Value must be a boolean"
                    ))
            elif definition.type == SettingType.NUMBER:
                if not isinstance(effective_value, (int, float)):
                    errors.append(SettingError(
                        identifier=identifier,
                        message="Value must be a number"
                    ))
                    continue
                
                assert isinstance(definition, NumericSetting)
                setting_uom = definition.uom
                provided_uom = provided.uom if provided is not None else None
                
                if provided_uom is not None and setting_uom is None:
                    errors.append(SettingError(
                        identifier=identifier,
                        message="Unit of measure is not supported for this setting"
                    ))
                    continue
                
                # Convert value if needed
                value_for_range = effective_value
                if (setting_uom is not None and 
                    provided_uom is not None and 
                    provided_uom != setting_uom):
                    converted = self._convert_uom(effective_value, provided_uom, setting_uom)
                    if converted is None:
                        errors.append(SettingError(
                            identifier=identifier,
                            message="Unit of measure is not convertible to required unit"
                        ))
                        continue
                    value_for_range = converted
                
                # Check range
                if definition.min_value is not None and value_for_range < definition.min_value:
                    errors.append(SettingError(
                        identifier=identifier,
                        message=f"Value must be >= {definition.min_value}"
                    ))
                if definition.max_value is not None and value_for_range > definition.max_value:
                    errors.append(SettingError(
                        identifier=identifier,
                        message=f"Value must be <= {definition.max_value}"
                    ))
        
        return errors
    
    @abstractmethod
    def apply_settings(self, settings: list[SettingValue]) -> list[SettingError]:
        """Apply settings to the machine.
        
        Args:
            settings: List of settings to apply
            
        Returns:
            List of errors (empty if successful)
        """
        pass
    
    def _convert_uom(
        self, 
        value: float, 
        from_uom: UnitOfMeasure, 
        to_uom: UnitOfMeasure
    ) -> Optional[float]:
        """Convert a value from one unit of measure to another.
        
        Args:
            value: The value to convert
            from_uom: The unit to convert from
            to_uom: The unit to convert to
            
        Returns:
            Converted value or None if conversion is not supported
        """
        if from_uom == to_uom:
            return value
        
        # Temperature conversions
        if from_uom == UnitOfMeasure.DEGREE_CELSIUS and to_uom == UnitOfMeasure.DEGREE_FAHRENHEIT:
            return (value * 9 / 5) + 32
        if from_uom == UnitOfMeasure.DEGREE_FAHRENHEIT and to_uom == UnitOfMeasure.DEGREE_CELSIUS:
            return (value - 32) * 5 / 9
        
        # Pressure conversions
        if from_uom == UnitOfMeasure.BAR and to_uom == UnitOfMeasure.PSI:
            return value * 14.5037738007218
        if from_uom == UnitOfMeasure.PSI and to_uom == UnitOfMeasure.BAR:
            return value / 14.5037738007218
        
        # Speed conversions
        if from_uom == UnitOfMeasure.RPM and to_uom == UnitOfMeasure.RPS:
            return value / 60
        if from_uom == UnitOfMeasure.RPS and to_uom == UnitOfMeasure.RPM:
            return value * 60
        
        # Time conversions
        if from_uom == UnitOfMeasure.SECOND and to_uom == UnitOfMeasure.MINUTE:
            return value / 60
        if from_uom == UnitOfMeasure.MINUTE and to_uom == UnitOfMeasure.SECOND:
            return value * 60
        
        return None

File: solution/TypeScript/packages/shared/src/providers/configurable.ts
export abstract class ConfigurableMachine {
    abstract get settings(): Setting[];

    protected verifySettings(settings: SettingValue[]): SettingError[] {
        // Perform the following checks:
        // * Each identifier must existing in settings (see getter)
        // * No duplicate identifiers
        // * There must be a value or a default value for each setting that is not nullable
        // * Value must be of the correct type
        // * If numeric, value must be >= minValue and <= maxValue
        // * If numeric, uom must fit or be convertible to the setting's uom
        const errors: SettingError[] = [];

        const definitionMap = new Map<string, Setting>();
        for (const def of this.settings) {
            // Machine definitions are assumed to be manually kept consistent/unique.
            definitionMap.set(def.identifier, def);
        }

        const seenProvided = new Set<string>();
        const duplicateProvided = new Set<string>();
        for (const provided of settings) {
            if (seenProvided.has(provided.identifier)) {
                duplicateProvided.add(provided.identifier);
            } else {
                seenProvided.add(provided.identifier);
            }
        }
        for (const id of duplicateProvided) {
            errors.push({ identifier: id, message: "Duplicate setting identifier provided" });
        }

        const providedMap = new Map<string, SettingValue>();
        for (const provided of settings) {
            if (!definitionMap.has(provided.identifier)) {
                errors.push({ identifier: provided.identifier, message: "Unknown setting identifier" });
                continue;
            }

            // Keep the first occurrence (duplicates already flagged).
            if (!providedMap.has(provided.identifier)) {
                providedMap.set(provided.identifier, provided);
            }
        }

        for (const [identifier, def] of definitionMap.entries()) {
            const provided = providedMap.get(identifier);
            const nullable = def.nullable === true;

            const hasProvidedValue = provided?.value !== undefined;
            const defaultValue = (def as NumericSetting | StringSetting | BooleanSetting).defaultValue;
            const hasDefaultValue = defaultValue !== undefined;

            if (!nullable && !hasProvidedValue && !hasDefaultValue) {
                errors.push({
                    identifier,
                    message: "Missing value (no default and not nullable)"
                });
                continue;
            }

            // If no value is provided, the default (if any) becomes effective.
            const effectiveValue = (hasProvidedValue ? provided?.value : defaultValue);
            if (effectiveValue === undefined) {
                // Nullable without value is fine.
                if (def.type !== SettingType.NUMBER && provided?.uom !== undefined) {
                    errors.push({
                        identifier,
                        message: "Unit of measure is only allowed for numeric settings"
                    });
                }
                continue;
            }

            if (def.type !== SettingType.NUMBER && provided?.uom !== undefined) {
                errors.push({
                    identifier,
                    message: "Unit of measure is only allowed for numeric settings"
                });
            }

            switch (def.type) {
                case SettingType.STRING:
                    if (typeof effectiveValue !== "string") {
                        errors.push({ identifier, message: "Value must be a string" });
                    }
                    break;
                case SettingType.BOOLEAN:
                    if (typeof effectiveValue !== "boolean") {
                        errors.push({ identifier, message: "Value must be a boolean" });
                    }
                    break;
                case SettingType.NUMBER: {
                    if (typeof effectiveValue !== "number" || Number.isNaN(effectiveValue)) {
                        errors.push({ identifier, message: "Value must be a number" });
                        break;
                    }

                    const settingUom = def.uom;
                    const providedUom = provided?.uom;

                    if (providedUom !== undefined && settingUom === undefined) {
                        errors.push({ identifier, message: "Unit of measure is not supported for this setting" });
                        break;
                    }

                    let valueForRange = effectiveValue;
                    if (
                        settingUom !== undefined &&
                        providedUom !== undefined &&
                        providedUom !== settingUom
                    ) {
                        const converted = this.convertUom(effectiveValue, providedUom, settingUom);
                        if (converted === undefined) {
                            errors.push({ identifier, message: "Unit of measure is not convertible to required unit" });
                            break;
                        }
                        valueForRange = converted;
                    }

                    if (def.minValue !== undefined && valueForRange < def.minValue) {
                        errors.push({ identifier, message: `Value must be >= ${def.minValue}` });
                    }
                    if (def.maxValue !== undefined && valueForRange > def.maxValue) {
                        errors.push({ identifier, message: `Value must be <= ${def.maxValue}` });
                    }
                    break;
                }
            }
        }

        return errors;
    }

    abstract applySettings(settings: SettingValue[]): SettingError[];

    protected convertUom(value: number, from: UnitOfMeasure, to: UnitOfMeasure): number | undefined {
        if (from === to) {
            return value;
        }

        // Note: don't use `switch ([from, to])` here - arrays compare by reference in JS/TS.
        if (from === UnitOfMeasure.DEGREE_CELSIUS && to === UnitOfMeasure.DEGREE_FAHRENHEIT) {
            return (value * 9 / 5) + 32;
        }
        if (from === UnitOfMeasure.DEGREE_FAHRENHEIT && to === UnitOfMeasure.DEGREE_CELSIUS) {
            return (value - 32) * 5 / 9;
        }
        if (from === UnitOfMeasure.BAR && to === UnitOfMeasure.PSI) {
            return value * 14.5037738007218;
        }
        if (from === UnitOfMeasure.PSI && to === UnitOfMeasure.BAR) {
            return value / 14.5037738007218;
        }
        if (from === UnitOfMeasure.RPM && to === UnitOfMeasure.RPS) {
            return value / 60;
        }
        if (from === UnitOfMeasure.RPS && to === UnitOfMeasure.RPM) {
            return value * 60;
        }
        if (from === UnitOfMeasure.SECOND && to === UnitOfMeasure.MINUTE) {
            return value / 60;
        }
        if (from === UnitOfMeasure.MINUTE && to === UnitOfMeasure.SECOND) {
            return value * 60;
        }

        return undefined;
    }
}

export enum SettingType {
    STRING = "string",
    NUMBER = "number",
    BOOLEAN = "boolean"
}

export enum UnitOfMeasure {
    DEGREE_CELSIUS = "°C",
    DEGREE_FAHRENHEIT = "°F",
    BAR = "bar",
    PSI = "psi",
    RPM = "rpm",
    RPS = "rps",
    SECOND = "s",
    MINUTE = "min",
}

type SettingBase = {
    namespace: string | undefined;
    identifier: string; // Must be globally unique
    description: string;
    nullable?: boolean | undefined;
}

export type NumericSetting = SettingBase & {
    type: SettingType.NUMBER;
    defaultValue?: number | undefined;
    uom?: UnitOfMeasure | undefined;
    minValue?: number | undefined;
    maxValue?: number | undefined;
}

export type StringSetting = SettingBase & {
    type: SettingType.STRING;
    defaultValue?: string | undefined;
    uom?: never;
}

export type BooleanSetting = SettingBase & {
    type: SettingType.BOOLEAN;
    defaultValue?: boolean | undefined;
    uom?: never;
}

export type Setting = NumericSetting | StringSetting | BooleanSetting;

export type SettingValue = {
    identifier: string; // Must be globally unique
    value: string | number | boolean | undefined;
    uom?: UnitOfMeasure | undefined;
}

export type SettingError = {
    identifier: string; // Must be globally unique
    message: string;
}
